## Classification of Weather Data using scikit-learn
In this notebook, we will use scikit-learn to perform a decision tree based classification of weather data.

**Importing necessary libraries**

In [1]:
import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier

In [3]:
from google.colab import files
uploaded = files.upload()

Saving daily_weather.csv to daily_weather.csv


Create a Pandas DataFrame to read from a CSV file

In [4]:
data = pd.read_csv("daily_weather.csv")
data.head()

number  air_pressure_9am  air_temp_9am  avg_wind_direction_9am  \
0       0        918.060000     74.822000              271.100000   
1       1        917.347688     71.403843              101.935179   
2       2        923.040000     60.638000               51.000000   
3       3        920.502751     70.138895              198.832133   
4       4        921.160000     44.294000              277.800000   

   avg_wind_speed_9am  max_wind_direction_9am  max_wind_speed_9am  \
0            2.080354              295.400000            2.863283   
1            2.443009              140.471548            3.533324   
2           17.067852               63.700000           22.100967   
3            4.337363              211.203341            5.190045   
4            1.856660              136.500000            2.863283   

   rain_accumulation_9am  rain_duration_9am  relative_humidity_9am  \
0                    0.0                0.0              42.420000   
1                    0.0                0.0              24.328697   
2                    0.0               20.0               8.900000   
3                    0.0                0.0              12.189102   
4                    8.9            14730.0              92.410000   

   relative_humidity_3pm  
0              36.160000  
1              19.426597  
2              14.460000  
3              12.742547  
4              76.740000

## Daily Weather Data Description


The file **daily_weather.csv** is a comma-separated file that contains weather data. This data comes from a weather station located in San Diego, California. The weather station is equipped with sensors that capture weather-related measurements such as air temperature, air pressure, and relative humidity. Data was collected for a period of three years, from September 2011 to September 2014, to ensure that sufficient data for different seasons and weather conditions is captured.


In [5]:
# Let's look at the columns in the dataset
data.columns

Index(['number', 'air_pressure_9am', 'air_temp_9am', 'avg_wind_direction_9am',
       'avg_wind_speed_9am', 'max_wind_direction_9am', 'max_wind_speed_9am',
       'rain_accumulation_9am', 'rain_duration_9am', 'relative_humidity_9am',
       'relative_humidity_3pm'],
      dtype='object')

<br>Each row in daily_weather.csv captures weather data for a separate day.  <br><br>
Sensor measurements from the weather station were captured at one-minute intervals.  These measurements were then processed to generate values to describe daily weather. Since this dataset was created to classify low-humidity days vs. non-low-humidity days (that is, days with normal or high humidity), the variables included are weather measurements in the morning, with one measurement, namely relatively humidity, in the afternoon.  The idea is to use the morning weather values to predict whether the day will be low-humidity or not based on the afternoon measurement of relative humidity.

Each row, or sample, consists of the following variables:

* **number:** unique number for each row
* **air_pressure_9am:** air pressure averaged over a period from 8:55am to 9:04am (*Unit: hectopascals*)
* **air_temp_9am:** air temperature averaged over a period from 8:55am to 9:04am (*Unit: degrees Fahrenheit*)
* **air_wind_direction_9am:** wind direction averaged over a period from 8:55am to 9:04am (*Unit: degrees, with 0 means coming from the North, and increasing clockwise*)
* **air_wind_speed_9am:** wind speed averaged over a period from 8:55am to 9:04am (*Unit: miles per hour*)
* ** max_wind_direction_9am:** wind gust direction averaged over a period from 8:55am to 9:10am (*Unit: degrees, with 0 being North and increasing clockwise*)
* **max_wind_speed_9am:** wind gust speed averaged over a period from 8:55am to 9:04am (*Unit: miles per hour*)
* **rain_accumulation_9am:** amount of rain accumulated in the 24 hours prior to 9am (*Unit: millimeters*)
* **rain_duration_9am:** amount of time rain was recorded in the 24 hours prior to 9am (*Unit: seconds*)
* **relative_humidity_9am:** relative humidity averaged over a period from 8:55am to 9:04am (*Unit: percent*)
* **relative_humidity_3pm:** relative humidity averaged over a period from 2:55pm to 3:04pm (*Unit: percent *)


### Data Cleaning

We have to look for any null values and get rid of those values to make a clean data set

In [6]:
data.isnull().any().any()

True

This means that there are sum null values. Let's have a look at the columns which have null values.

In [7]:
data.isnull().sum()

number                    0
air_pressure_9am          3
air_temp_9am              5
avg_wind_direction_9am    4
avg_wind_speed_9am        3
max_wind_direction_9am    3
max_wind_speed_9am        4
rain_accumulation_9am     6
rain_duration_9am         3
relative_humidity_9am     0
relative_humidity_3pm     0
dtype: int64

In [8]:
# Print the rows with missing values
data[data.isnull().any(axis = 1)]

number  air_pressure_9am  air_temp_9am  avg_wind_direction_9am  \
16        16        917.890000           NaN              169.200000   
111      111        915.290000     58.820000              182.600000   
177      177        915.900000           NaN              183.300000   
262      262        923.596607     58.380598               47.737753   
277      277        920.480000     62.600000              194.400000   
334      334        916.230000     75.740000              149.100000   
358      358        917.440000     58.514000               55.100000   
361      361        920.444946     65.801845               49.823346   
381      381        918.480000     66.542000               90.900000   
409      409               NaN     67.853833               65.880616   
517      517        920.570000     53.600000              100.100000   
519      519        916.250000     55.670000              176.400000   
546      546               NaN     42.746000              251.100000   
620      620        921.200000     56.786000              192.300000   
625      625        912.400000     50.774000              171.600000   
656      656        920.830000     66.344000                     NaN   
670      670        910.920000     48.362000              156.500000   
672      672        922.448945     72.863773                     NaN   
705      705        911.900000     59.072000              199.800000   
731      731        922.970166     51.391847               33.810942   
737      737        917.895130     76.804690              104.771020   
788      788        917.923442     73.249717               42.101739   
840      840        918.043767           NaN              181.774042   
848      848        915.250000     37.562000              246.500000   
861      861        919.065408           NaN              172.303728   
869      869               NaN     45.104000              259.000000   
998      998        914.140000     71.240000                     NaN   
1031    1031        922.669195           NaN               47.946284   
1035    1035        919.670000     77.576000              171.800000   
1063    1063        917.300185     65.790001                     NaN   
1066    1066        919.564869     73.726732               68.704694   

      avg_wind_speed_9am  max_wind_direction_9am  max_wind_speed_9am  \
16              2.192201              196.800000            2.930391   
111            15.613841              189.000000                 NaN   
177             4.719943              189.900000            5.346287   
262            10.636273               67.145843           13.671423   
277             2.751436                     NaN            3.869906   
334             2.751436              187.500000            4.183078   
358            10.021491                     NaN           12.705819   
361            21.520177               61.886944           25.549112   
381             3.467257               89.400000            4.406772   
409             4.328594               78.570923            5.216734   
517             4.697574                     NaN            6.285801   
519             6.666081              188.200000                 NaN   
546            12.929513              274.400000           17.604718   
620             9.551734              201.400000           11.005745   
625                  NaN              181.400000            4.831790   
656            15.457255              189.400000           16.486248   
670                  NaN              177.500000           16.128337   
672             3.682370              214.196160            4.849450   
705             1.275056              239.500000            1.834291   
731                  NaN               59.290089           11.111555   
737             1.632705               97.178763                 NaN   
788             4.132698               64.284969            5.345258   
840             0.964376              185.618

#### Data Cleaning Steps

In [9]:
# We do not need to number the rows as Pandas provides its's own indexing
del data['number']
data.columns

Index(['air_pressure_9am', 'air_temp_9am', 'avg_wind_direction_9am',
       'avg_wind_speed_9am', 'max_wind_direction_9am', 'max_wind_speed_9am',
       'rain_accumulation_9am', 'rain_duration_9am', 'relative_humidity_9am',
       'relative_humidity_3pm'],
      dtype='object')

Next, we drop out the rows with null values

In [10]:
before_rows = data.shape[0]
data = data.dropna()
after_rows = data.shape[0]

In [ ]:
print("The number of dropped rows are {}".format(before_rows - after_rows))

The number of dropped rows are 31


## Convert to a Classification task

**Binarize the relative humidity_3pm to 0 or 1**

We are assigning the values 0 or 1 and adding a new column 'high humidity label'. We are basically classifying the data into two categories ( binary problem ) by setting a desired value ( 24.99 , in this case ) to be the threshold and anything above is high ( 1 ) and anything below is low ( 0 ).

In [11]:
clean_data = data.copy() # New data frame to avoid confusion
clean_data['high_humidity_label'] = (clean_data['relative_humidity_3pm'] > 24.99) * 1
print(clean_data['high_humidity_label'])

0       1
1       0
2       0
3       0
4       1
       ..
1090    1
1091    1
1092    1
1093    1
1094    0
Name: high_humidity_label, Length: 1064, dtype: int64


Target is now stored as y. Here, target is the label - 'high_humidity_label'

In [12]:
y = clean_data[['high_humidity_label']].copy()
y

high_humidity_label
0                       1
1                       0
2                       0
3                       0
4                       1
...                   ...
1090                    1
1091                    1
1092                    1
1093                    1
1094                    0

[1064 rows x 1 columns]

In [13]:
clean_data['relative_humidity_3pm'].head()

0    36.160000
1    19.426597
2    14.460000
3    12.742547
4    76.740000
Name: relative_humidity_3pm, dtype: float64

In [14]:
y.head()

high_humidity_label
0                    1
1                    0
2                    0
3                    0
4                    1

### Use 9am Sensor signals to predict Humidity at 3PM

In [15]:
time = '9am'
features = list(clean_data.columns[clean_data.columns.str.contains(time)])

# we do not need relative humidity at 9am
features.remove('relative_humidity_9am')

features

['air_pressure_9am',
 'air_temp_9am',
 'avg_wind_direction_9am',
 'avg_wind_speed_9am',
 'max_wind_direction_9am',
 'max_wind_speed_9am',
 'rain_accumulation_9am',
 'rain_duration_9am']

In [16]:
# Make the data of these features as X
X = clean_data[features].copy()
#X

## Perform the test and Train split

### REMINDER: Training Phase

* In the **training phase**, the learning algorithm uses the training data to adjust the model’s parameters to minimize errors.  At the end of the training phase, you get the trained model.


* In the **testing phase**, the trained model is applied to test data.  Test data is separate from the training data, and is previously unseen by the model.  The model is then evaluated on how it performs on the test data.  The goal in building a classifier model is to have the model perform well on training as well as test data.


In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.33, random_state = 324)

Let us look at these sets using the following commands below.

In [ ]:
# type(X_train)
# type(X_test)
# type(y_train)
# type(y_test)
# X_train.head()
# #y_train.describe()

In [18]:
y_train.describe()


high_humidity_label
count           712.000000
mean              0.494382
std               0.500320
min               0.000000
25%               0.000000
50%               0.000000
75%               1.000000
max               1.000000

In [19]:
X_train.describe()

air_pressure_9am  air_temp_9am  avg_wind_direction_9am  \
count        712.000000    712.000000              712.000000   
mean         918.913897     65.194366              142.123330   
std            3.147923     11.210412               68.773699   
min          907.990000     36.752000               15.500000   
25%          916.727792     57.375500               65.282862   
50%          919.000000     65.967556              166.225018   
75%          921.134993     73.908500              190.500000   
max          929.320000     91.112000              343.400000   

       avg_wind_speed_9am  max_wind_direction_9am  max_wind_speed_9am  \
count          712.000000              712.000000          712.000000   
mean             5.568288              147.789373            7.089629   
std              4.467828               67.186609            5.486311   
min              0.782929               31.800000            1.185578   
25%              2.304048               75.978656            3.168182   
50%              3.958486              176.800000            5.077854   
75%              7.504934              201.400000            9.104346   
max             21.541732              299.200000           26.351153   

       rain_accumulation_9am  rain_duration_9am  
count             712.000000         712.000000  
mean                0.202399         282.884615  
std                 1.628988        1584.404987  
min                 0.000000           0.000000  
25%                 0.000000           0.000000  
50%                 0.000000           0.000000  
75%                 0.000000           0.000000  
max                24.020000       17704.000000

### Fit the model on the training set

We will build a model using the Decision Tree Classifier using the fit functions.

In [20]:
humidity_classifier = DecisionTreeClassifier(max_leaf_nodes = 10, random_state = 0)
humidity_classifier.fit(X_train, y_train)

DecisionTreeClassifier(max_leaf_nodes=10, random_state=0)

In [21]:
type(humidity_classifier)

sklearn.tree._classes.DecisionTreeClassifier

### Test the model on thetesting set



In [22]:
predictions = humidity_classifier.predict(X_test)
type(predictions)

numpy.ndarray

In [23]:
predictions[:10]
#predictions[:len(predictions)]

array([0, 0, 1, 1, 1, 1, 0, 0, 0, 1])

In [24]:
y_test[['high_humidity_label']][:10]


high_humidity_label
456                     0
845                     0
693                     1
259                     1
723                     1
224                     1
300                     1
442                     0
585                     1
1057                    1

So, we have the actual values in the y_test result and the predicted values in the predictions set and now we can compare all the predictions with the actual results to see if we were correct or not. Let's find out the accuracy using accuracy_score.

### Measure the accuracy of the data

In [25]:
accuracy_score(y_test, y_pred = predictions)

0.8153409090909091

### Measuring the mean squared error

In [26]:
from sklearn.metrics import mean_squared_error
mean_squared_error(y_test, y_pred = predictions)

0.1846590909090909

We have predictd the humidity at 3PM based on the 9AM measurements with an 81% accuracy and 19% loss which are very good stats. Hence, it's a success.